File: atoi.py
#!/usr/bin/env python
# encoding: utf8

# String to Integer (atoi): Implement atoi to convert a string to an integer. 

__author__ = "Zhu Xianfeng <xianfeng.zhu@gmail.com>"

class Solution:
    # @return an integer
    def atoi(self, s):
        INT_MAX = 2 ** 31 - 1
        INT_MIN = -(INT_MAX + 1)
        slen = len(s)
        if slen == 0:
            return 0

        # Ignore blank
        i = 0
        while s[i] in (' ', '\t', '\n', '\0'):
            i += 1
        if i != 0:
            s = s[i:]

        mlen = str(INT_MIN)
        slen = len(s)
        if slen > mlen or slen == 0:
            return 0

        # Handle flag
        flag = 1
        if s[0] == '+':
            s = s[1:]
        elif s[0] == '-':
            flag = -1
            s = s[1:]

        n = 0
        for i in s:
            if i not in "0123456789":
                break
            i = int(i)
            n = n * 10 + i

        n = (n * flag)
        if n > INT_MAX:
            n = INT_MAX
        elif n < INT_MIN:
            n = INT_MIN
        return n

def main():
    s = "2147483648"
    sol = Solution()
    print sol.atoi(s)

if __name__ == "__main__":
    main()

File: algorithms/atoi.cpp
// 8. String to Integer (atoi)b)
// Author: xianfeng.zhu@gmail.com

#include <ctype.h>
#include <stdio.h>
#include <limits.h>
#include <string>

class Solution
{
public:
    int myAtoi(const std::string& str)
    {
        int num = 0;
        int idx = 0;

        // Ignore whitespace character
        while (idx < str.size() && str[idx] == ' ')
        {
            idx++;
        }
        if (idx == str.size())
        {
            return 0;
        }

        // Determine minus sign
        bool minus = false;
        if (str[idx] == '-' || str[idx] == '+')
        {
            minus = (str[idx] == '-' ? true : false);
            idx++;
        }

        while (idx < str.size() && isdigit(str[idx]) != 0)
        {
            int digit = str[idx] - '0';
            digit *= (minus ? -1 : 1);
            if ((num > INT_MAX / 10) || (num == INT_MAX / 10 && digit >= INT_MAX % 10))
            {
                // Overflow, bigger than INT_MAX
                return INT_MAX;
            }
            if ((num < INT_MIN / 10) || (num == INT_MIN / 10 && digit <= INT_MIN % 10))
            {
                // Overflow, smaller than INT_MIN
                return INT_MIN;
            }

            num = num * 10 + digit;
            idx++;
        }

        return num;
    }
};

int main(int argc, char* argv[])
{
    std::string str = "-91283472332";
    if (argc > 1)
    {
        str = argv[1];
    }
    int num = Solution().myAtoi(str);
    printf("Input: %s\n", str.c_str());
    printf("Output: %d\n", num);
    return 0;
}
